samcli/lib/build/build_graph.py

```diff
@@ -0,0 +1,188 @@
+        are no longer used, and they can be deleted
+
+        If persist parameter is given True, build graph is written to .aws-sam/build.toml file
+        """
+        self._build_definitions[:] = [bd for bd in self._build_definitions if len(bd.functions) > 0]
+        if persist:
+            self._write()
+
+    def _read(self):
+        """
+        Reads build.toml file into array of build definition
+        Each build definition will have empty function list, which will be populated from the current template.yaml file
+        """
+        if not self._build_definitions:
+            LOG.debug("Instantiating build definitions")
+            self._build_definitions = []
+            document = {}
+            try:
+                txt = self._filepath.read_text()
+                document = tomlkit.loads(txt)
+            except OSError:
+                LOG.debug("No previous build graph found, generating new one")
+            build_definitions_table = document.get(BuildGraph.BUILD_DEFINITIONS, [])
+            for build_definition_key in build_definitions_table:
+                build_definition = _toml_table_to_build_definition(build_definition_key,
+                                                                   build_definitions_table[build_definition_key])
+                self._build_definitions.append(build_definition)
+
+        return self._build_definitions
+
+    def _write(self):
+        """
+        Writes build definition details into build.toml file, which would be used by the next build.
+        build.toml file will contain the same information as build graph,
+        function details will only be preserved as function names
+        """
+        # convert build definition list into toml table
+        build_definitions_table = tomlkit.table()
+        for build_definition in self._build_definitions:
+            build_definition_as_table = _build_definition_to_toml_table(build_definition)
+            build_definitions_table.add(build_definition.uuid, build_definition_as_table)
+
+        # create toml document and add build definitions
+        document = tomlkit.document()
+        document.add(tomlkit.comment("This file is auto generated by SAM CLI build command"))
+        document.add(BuildGraph.BUILD_DEFINITIONS, build_definitions_table)
+
+        if not self._filepath.exists():
+            open(self._filepath, "a+").close()
+
+        self._filepath.write_text(tomlkit.dumps(document))
+
+
+class BuildDefinition:
+    """
+    Build definition holds information about each unique build
+    """
+
+    def __init__(self, runtime, codeuri, metadata):
+        self.runtime = runtime
+        self.codeuri = codeuri
+        self.metadata = metadata if metadata else {}
+        self.uuid = str(uuid4())
+        self.functions = []
+
+    def add_function(self, function):
+        self.functions.append(function)
+
+    def get_function_name(self):
+        return self.functions[0].name
```

Comment:
Check in case functions is empty.

Reply:
Build process is calling `clean_redundant_functions_and_update` method to clean any build definitions without functions in it, but I will add secondary check here in case something goes bad in the future.